experimental/oak_baremetal_kernel/src/mm/bitmap_frame_allocator.rs

```diff
@@ -0,0 +1,336 @@
+/// memory with a bitmap.
+///
+/// This implementation tracks memory using two bitmaps:
+///   - `valid`, which tracks whether a frame can be allocated at all (not all physical frames are
+///     usable memory)
+///   - `allocated`, which tracks whether a frame is currently allocated or not.
+///
+/// A frame can be in one of three states:
+///  valid = 0, allocated = _ -- the allocator will never hand out said frame.
+///  valid = 1, allocated = 0 -- the frame is eligible for allocation.
+///  valid = 1, allocated = 1 -- the frame is currently in use.
+///
+/// The BitmapAllocator itself does not use memory allocation (no heap is required), but rather
+/// expects that the const parameter N is set to the exact number of u64-s required to construct a
+/// bitmap covering the memory range.
+/// (In container-ish terms, N specifies the capacity of the allocator, which can not be resized.)
+#[derive(Debug)]
+pub(crate) struct BitmapAllocator<S: PageSize, const N: usize> {
+    allocated: BitArray<[u64; N], Lsb0>,
+    valid: BitArray<[u64; N], Lsb0>,
+    range: PhysFrameRangeInclusive<S>,
+}
+
+impl<S: PageSize, const N: usize> BitmapAllocator<S, N> {
+    /// Creates a new bitmap allocator for a physical frame range.
+    /// Panics if N does not match the number of u64-s required to track all frames in that range.
+    /// Initially, the allocator will mark the whole range as invalid.
+    #[allow(dead_code)]
+    pub fn new(range: PhysFrameRangeInclusive<S>) -> Self {
+        // Unfortunately there doesn't seem to be a way to hoist this to the type system.
+        let expected = bitvec::mem::elts::<u64>(range.count());
+        if expected != N {
+            panic!(
+                "BitmapAllocator bitmap size does not match FrameRange size; expected {}, got {}",
+                expected, N
+            );
+        }
+
+        Self {
+            allocated: BitArray::ZERO,
+            valid: BitArray::ZERO,
+            range,
+        }
+    }
+
+    /// Marks a region of memory as either valid or invalid.
+    ///
+    /// Allocations can happen only from regions that are valid. This method does not check whether
+    /// any allocations have been made from regions to be marked as invalid, and panics if the range
+    /// is outside the range of the allocator.
+    #[allow(dead_code)]
+    pub fn mark_valid(&mut self, range: PhysFrameRangeInclusive<S>, valid: bool) {
+        if let (Some(start), Some(end)) = (self.frame_idx(range.start), self.frame_idx(range.end)) {
+            self.valid.get_mut(start..end + 1).unwrap().fill(valid);
+        } else {
+            panic!(
+                "Can't mark validity for frame range that's outside our range; our range: {:?}, frame range: {:?}",
+                self.range, range
+            );
+        }
+    }
+
+    /// Returns the largest contiguous section of unallocated memory.
+    #[allow(dead_code)]
+    pub fn largest_available(&self) -> Option<PhysFrameRangeInclusive<S>> {
```

Comment:
I am not sure whether this is a particularly useful function, as I think we would likely want to know the first location where a range or a specific size can fit, rather than overall the largest one.

Reply:
Longer-term, you are correct: we probably need something that will allow us to allocate a contiguous range of physical memory, and won't care about its exact location.

I implemented this method to be compatible with our existing heap allocation mechanism, where we just take the biggest slice of memory available and use it as the heap. In the future we won't want to give _all_ the memory to the kernel heap, at which point we can get rid of this function.